### PR TITLE
fix: Ignore PAT with emtpy SCM url

### PR DESCRIPTION
**Pull request by tolusha**
### What does this PR do?
Ignore PAT with emtpy SCM url

### Screenshot/screencast of this PR

N/A

### What issues does this PR fix or reference?

https://issues.redhat.com/browse/CRW-5291

### How to test this PR?

N/A

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

**Review by ibuziuk — COMMENTED** [review]
@tolusha @vinokurig for me personally it is very hard to grok the logic in checks. Would it be possible to do some refactoring (extra private methods for checks?) + add unit tests?

**Review by amisevsk — APPROVED** [review]
LGTM but I agree that the overall logic could be made more readable (especially with nested conditions)

File: infrastructures/infrastructure-factory/src/main/java/org/eclipse/che/api/factory/server/scm/kubernetes/KubernetesPersonalAccessTokenManager.java [hunk 1]
```diff
@@ -186,6 +188,7 @@ private Optional<PersonalAccessToken> doGetPersonalAccessToken(
               && (oAuthProviderName == null
                   || oAuthProviderName.equals(
                       annotations.get(ANNOTATION_SCM_PERSONAL_ACCESS_TOKEN_NAME)))
+              && (!isNullOrEmpty(trimmedUrl)) // ignore PAT with empty server endpoint URL
```
Comment (vinokurig) [inline infrastructures/infrastructure-factory/src/main/java/org/eclipse/che/api/factory/server/scm/kubernetes/KubernetesPersonalAccessTokenManager.java:191]: The `ANNOTATION_SCM_URL` is mandatory so I think it is better to move this check to the beginning of the loop body: 
```java
if (isNullOrEmpty(trimmedUrl)) {
   continue;
}
```

**Review by vinokurig — APPROVED** [review]
Tested by adding and changing a PAT from GitHub, works as expected.
@tolusha how is this change related to the Jira issue?

**Comment by tolusha** [comment]
We had the bug on a dashboard side, when PAT is created for BitBucket with an empty scm url.
It lead to the problem when che-server failed and it was not possible to log in anymore.
So, we fixed dashboard [1] and also protected che-serve with this PR.
[1] https://github.com/eclipse-che/che-dashboard/pull/991

**Pull request merged**